homeassistant/components/google_mail/sensor.py

```diff
@@ -0,0 +1,156 @@
+"""Support for Google Mail Sensors."""
+from __future__ import annotations
+
+from datetime import datetime, timedelta, timezone
+from typing import Any
+
+from google.auth.exceptions import RefreshError
+from google.oauth2.credentials import Credentials
+from googleapiclient.discovery import build
+import voluptuous as vol
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_TOKEN
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.config_entry_oauth2_flow import OAuth2Session
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.service import async_extract_config_entry_ids
+
+from .const import (
+    ATTR_ENABLED,
+    ATTR_END,
+    ATTR_MESSAGE,
+    ATTR_PLAIN_TEXT,
+    ATTR_RESTRICT_CONTACTS,
+    ATTR_RESTRICT_DOMAIN,
+    ATTR_START,
+    ATTR_TITLE,
+    DOMAIN,
+)
+from .entity import GoogleMailEntity
+
+SCAN_INTERVAL = timedelta(minutes=15)
+
+SERVICE_SET_VACATION = "set_vacation"
+
+SERVICE_SCHEMA = vol.All(
+    cv.make_entity_service_schema(
+        {
+            vol.Required(ATTR_ENABLED, default=True): cv.boolean,
+            vol.Optional(ATTR_TITLE): cv.string,
+            vol.Required(ATTR_MESSAGE): cv.string,
+            vol.Optional(ATTR_PLAIN_TEXT, default=True): cv.boolean,
+            vol.Optional(ATTR_RESTRICT_CONTACTS): cv.boolean,
+            vol.Optional(ATTR_RESTRICT_DOMAIN): cv.boolean,
+            vol.Optional(ATTR_START): cv.date,
+            vol.Optional(ATTR_END): cv.date,
+        },
+    )
+)
+
+SENSOR_TYPE = SensorEntityDescription(
+    key="vacation_end_date",
+    name="Vacation end date",
+    icon="mdi:clock",
+    device_class=SensorDeviceClass.TIMESTAMP,
+)
+
+
+async def async_setup_service(hass: HomeAssistant) -> None:
+    """Add the services for Google Mail."""
+
+    async def extract_gmail_config_entries(call: ServiceCall) -> list[ConfigEntry]:
+        return [
+            entry
+            for entry_id in await async_extract_config_entry_ids(hass, call)
+            if (entry := hass.config_entries.async_get_entry(entry_id))
+            and entry.domain == DOMAIN
+        ]
+
+    def _set_vacation(call: ServiceCall, entry: ConfigEntry) -> None:
+        """Run vacation call in the executor."""
+        credentials = Credentials(entry.data[CONF_TOKEN][CONF_ACCESS_TOKEN])
```

Comment:
While I think what is here is totally is equivalent, getting the token from `session.token` seems like it may be a bit a little more consistent with the API design to not have to dig into the config entry details. e.g. first call `await session.async_ensure_token_valid()` then `session.token` to get the token that has been ensured has been valid.  I think this lets you avoid needing to keep around the config entry in `GoogleMailEntity` since you already have to get the session anyway.